dali/core/mm/default_resources.cc

```diff
@@ -0,0 +1,282 @@
+  return CallAtExit<Callable>(std::forward<Callable>(c));
+}
+
+inline std::shared_ptr<device_async_resource> CreateDefaultDeviceResource() {
+  static CUDARTLoader CUDAInit;
+#ifdef DALI_USE_RMM_DEVICE_RESOURCE
+  auto upstream = std::make_shared<rmm::mr::cuda_memory_resource>();
+  return rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(std::move(upstream));
+#else
+  using resource_type = mm::async_pool_resource<mm::memory_kind::device>;
+  static auto upstream = std::make_shared<mm::cuda_malloc_memory_resource>();
+  auto rsrc = std::make_shared<resource_type>(upstream.get());
+  return make_shared_composite_resource(std::move(rsrc), upstream);
+#endif
+}
+
+inline std::shared_ptr<pinned_async_resource> CreateDefaultPinnedResource() {
+  using resource_type = mm::async_pool_resource<mm::memory_kind::pinned>;
+  static auto upstream = std::make_shared<rmm::mr::pinned_memory_resource>();
+  auto rsrc = std::make_shared<resource_type>(upstream.get());
+  return make_shared_composite_resource(std::move(rsrc), upstream);
+}
+
+template <memory_kind kind>
+const std::shared_ptr<default_memory_resource_t<kind>> &ShareDefaultResourceImpl();
+
+template <>
+const std::shared_ptr<host_memory_resource> &ShareDefaultResourceImpl<memory_kind::host>() {
+  if (!g_resources.host)
+    g_resources.host = CreateDefaultHostResource();
+  return g_resources.host;
+}
+
+template <>
+const std::shared_ptr<pinned_async_resource> &ShareDefaultResourceImpl<memory_kind::pinned>() {
+  if (!g_resources.pinned_async) {
+    static CUDARTLoader init_cuda;  // force initialization of CUDA before creating the resource
+    static auto cleanup = AtExit([] {
+      g_resources.ReleasePinned();
+    });
+    g_resources.pinned_async = CreateDefaultPinnedResource();
+  }
+  return g_resources.pinned_async;
+}
+
+const std::shared_ptr<device_async_resource> &ShareDefaultDeviceResourceImpl(int device_id) {
+  if (device_id < 0) {
+    CUDA_CALL(cudaGetDevice(&device_id));
+  }
+  if (g_resources.device.empty()) {
+    int ndevs = 0;
+    CUDA_CALL(cudaGetDeviceCount(&ndevs));
+    g_resources.device.resize(ndevs);
+  }
+  if (static_cast<size_t>(device_id) >= g_resources.device.size())
+    throw std::out_of_range(make_string(device_id, " is not a valid CUDA device index."));
+  if (!g_resources.device[device_id]) {
+    DeviceGuard devg(device_id);
+    static CUDARTLoader init_cuda;  // force initialization of CUDA before creating the resource
+    static auto cleanup = AtExit([] {
+      g_resources.ReleaseDevice();
+    });
+    g_resources.device[device_id] = CreateDefaultDeviceResource();
+  }
+  return g_resources.device[device_id];
+}
+
+template <>
+const std::shared_ptr<device_async_resource> &ShareDefaultResourceImpl<memory_kind::device>() {
+  return ShareDefaultDeviceResourceImpl(-1);
+}
+
+}  // namespace
+
+
+template <> DLL_PUBLIC
+void SetDefaultResource<memory_kind::host>(shared_ptr<host_memory_resource> resource) {
```

Comment:
nitpick: Sometimes there is std::shared_ptr and others shared_ptr in this file. I'd unify.